# Conceptual Questions: Programming for Analytics

**Keith Faunce**

_No LLMs were used in the coding process, notes & web search were viable to use_

### Q1. Assume that there exists a string called my_str. Write a program to check if there are duplicate characters (i.e., if any character appears more than once) in my_str. Print out the characters that are duplicate, along with their respective duplicate counts. In your output, make sure that each duplicate character (if any) is printed out only a single time.

In [2]:
import string
alphabet_str = string.ascii_lowercase
#print(alphabet_str)
i = 0
my_str = str(input("Please enter a sentence."))
lower_sent = my_str.lower()

for i in range(len(alphabet_str)):
    letter = alphabet_str[i]
    num = lower_sent.count(letter)
    if num > 1:
        print(f"{letter} has {num} duplicates")



Please enter a sentence. what duplicates arrre in this sentence?


a has 3 duplicates
c has 2 duplicates
e has 5 duplicates
h has 2 duplicates
i has 3 duplicates
n has 3 duplicates
r has 3 duplicates
s has 3 duplicates
t has 4 duplicates


### Q2. Assume that there exists a file named “aFile.txt” in the local disk. Write a program that prompts the user for a positive integer n as an input. The program should write the first n characters of each line in “aFile.txt” to a new file, named “newFile.txt”. The program should write the entire line, if n is greater than that line’s length. Below is a sample input and output for n = 10.



In [1]:
line = str("This is the line that will be tested")
while True:
    num = int(input("Enter a positive integer"))
    if num < 0:
        continue
    else:
        break
#aFile = open('C:/Users/Keith/MS_BA-Sem1/ism_sum/txts/aFile.txt', 'r')
#lines = aFile.readlines()
lines = [line.rstrip() for line in open('C:/Users/Keith/MS_BA-Sem1/ism_sum/txts/aFile.txt')] # had to remove trailing /n from readlines() results
#print(lines)
newFile = open('C:/Users/Keith/MS_BA-Sem1/ism_sum/txts/newFile.txt', 'w')
for element in lines:
    if num >= len(element):
        newFile.write(element)
        newFile.write('\n') # to add a newline, if not text stacks ontop like end =''
    else:
        newFile.write(element[:num])
        newFile.write('\n')

newFile.close()

Enter a positive integer 10


### Q3. Assume that there exists a file named “numbersWords.txt” in the local disk. Each line of this file consists of some integer numbers (int) and words (str). Write a program that finds the largest number in this file, and prints out this number.


In [3]:
import numpy as np

numWord = open('C:/Users/Keith/MS_BA-Sem1/ism_sum/txts/numbersWords.txt', 'r')
blob = numWord.read()
#print(blob)
data = blob.split()
#arr = np.array((data), dtype='float32')
max_num = 0
for element in data:
    if element.isdigit():
        i_element = int(element)
        if i_element > max_num:
            max_num = i_element
        

print(max_num)

15663


### Q4. Yoda Speech generator: Given a sentence string that consists of multiple words, divide the sentence into two parts and re-order the parts into a new string. If the sentence contains an even number of words (e.g., 2n), then each part should contain the same number of words (e.g., n for each part). If the sentence contains an odd number of words, then the 1st part in the original sentence should contain one less word then the 2nd part in the original sentence (e.g., n-1 for the 1st part and n for the 2nd part).


In [4]:
yoda = str(input("Enter a sentence"))
yoda_list = yoda.split()
#print(yoda_list)
total = ''
i = 0


if len(yoda_list) % 2 == 0:
    space = len(yoda_list) // 2
    for i in range(space, len(yoda_list)):
        total += yoda_list[space]
        total += ' '
        space += 1
    for i in range(0, len(yoda_list) // 2):
        total += yoda_list[i]
        total += ' '
        i += 1
else:
    space = len(yoda_list) // 2
    #space = space - 1
    for i in range(space, len(yoda_list)):
        total += yoda_list[space]
        total += ' '
        space += 1
    for i in range(0, len(yoda_list) // 2):
        total += yoda_list[i]
        total += ' '
        i += 1

print(total)

Enter a sentence Anakin you must not fall to the dark side


fall to the dark side Anakin you must not 


### Q5. Write a program that prompts the user to enter a 5 digit number such that the digits are in ascending order (including equality). Two valid examples are: 12568 and 11225. The program should stop and congratulate the user if the number has 5 digits and the digits are in ascending order. If the user does not enter a 5 digit number or enters a 5 digit number that is not in ascending order, then it should print out a warning message (with the corresponding problem) and should prompt for the input one more time. The program should loop until the user enters 5 digits that are in order.


In [5]:
#x = int(input("Enter a 5 digit ASC num. Ex: 12568, 11225"))
#y = str(x)
flagStaff = True
i = 0
j = 1

while flagStaff:
    x = int(input("Enter a 5 digit ASC num. Ex: 12568, 11225: "))
    #modify = str(x) + '9'
    y = str(x)
    if len(y) != 5:
        print("Please make sure the number is 5 digits")
        continue
    if y[0] > y[1] or y[1] > y[2] or y[2] > y[3] or y[3] > y[4]:
        print("Please make sure the number is in ASC order")
        continue
    else:
        flagStaff = False


# I think I was very close to a cleaner loop, I couldn't break out of the for loop while weighing the number values against one another
# without cancelling the full loop
    #for i in range(len(y)):
        #if y[i] > modify[j]:
            #print("Please make sure the number is in ASC order")
            #break
        #elif y[i] <= modify[j]:
            #flagStaff = False

        #i += 1
        #j += 1
        
print("Congratulations! Your number is in correct format")

Enter a 5 digit ASC num. Ex: 12568, 11225:  67584


Please make sure the number is in ASC order


Enter a 5 digit ASC num. Ex: 12568, 11225:  44636


Please make sure the number is in ASC order


Enter a 5 digit ASC num. Ex: 12568, 11225:  12345


Congratulations! Your number is in correct format


### Q6. Write a program that: 1) reports the attendance count of each unique student, and 2) prints out the student that had the highest attendance along with his/her attendance count. An example output is as follows:

**Output:**

Joe || Attendance count: 3

Bill || Attendance count: 2

Sarah || Attendance count: 2

Fred || Attendance count: 2

--------------------------
Highest attendance count: 3 || Student: Joe

In [7]:
import pandas as pd

day = pd.read_csv(r"C:\Users\Keith\MS_BA-Sem1\ism_sum\txts\attendance.txt")
#day
students = day.iloc[:, 1:].values.flatten()
students = pd.Series(students).dropna()
students
total_each = students.value_counts()
#total_each
big = 0
total_max = total_each.max()
for name, total in total_each.items():
    print(f"{name} || Attendance count: {total}")

print("-----------------------")
most = total_max
best_student = total_each[total_each == most].index.tolist()
for value in best_student:
    print(f"Highest attendance count: {most} || Student: {value}")

 Joe || Attendance count: 3
 Bill || Attendance count: 2
 Sarah || Attendance count: 2
 Fred || Attendance count: 2
-----------------------
Highest attendance count: 3 || Student:  Joe


### Q7. Suppose that you are the manager of a manufacturing firm who is trying to determine the optimal price of a product based on demand. The relationship between price and demand is as follows: 
Demand = 800 – 10 x Price
#### Each product costs 50 USD to manufacture. The profit of the company is calculated as: 
Profit = (Price – Cost per item) x Demand

Write a function that, given a single price input, will first calculate its demand and then return the profit amount. Then, search a range of prices (using the function above) to identify the optimal price that will provide you the maximum profit. Notice that the range is limited, because the demand goes negative if price exceeds a certain threshold. Print out this price and the maximum profit amount.


In [8]:
def information(price):
    per_item = 50
    demand = 800 - 10 * price
    profit = (price - per_item) * demand
    return profit

def main():
    max_profit = 0
    max_price = 0
    for price in range(0, 81):
        profit = information(price)
        if profit > max_profit:
            max_profit = profit
            best_price = price

    print(f"The best price is {best_price} with the most profit of {max_profit}")

if __name__ == "__main__":
    main()

The best price is 65 with the most profit of 2250


### Q8. Write a program that prompts the user for a number that is more than two digits (e.g., 123, 1234, 12345, etc.). The program should divide this input into three parts (according to the rules below) and then print them as a single string joined by hyphens (-). 

- If the number of digits is divisible by 3, then each part should have equal number of digits. For instance, if the input is "123", then the output should be: “1-2-3”. 
- If the number of digits is not divisible by 3, then the first two parts should have equal number of digits whereas the last part should be longer than the first two. For instance, if the input is "54321", then the result should be "5-4-321". If the input is "1234567", then the output should be "12-34-567".

In [1]:
num = (input("Enter a number longer than two digits"))


if len(num) % 3 == 0:
    space = len(num) // 3
    duo = space * 2
    p1 = num[:space]
    p2 = num[space:duo]
    p3 = num[duo:]
    print(p1 + "-" + p2 + "-" + p3)
else:
    space = len(num) // 3
    duo = space * 2
    p1 = num[:space]
    p2 = num[space:duo]
    p3 = num[duo:]
    print(p1 + "-" + p2 + "-" + p3)

Enter a number longer than two digits 12345678


12-34-5678
